File: 卡码网/Python/test23.py
#23. 二叉树的高度
import sys

class Node:
	def __init__(self,data):
		self.data=data
		self.left=None
		self.right=None
	def getHeight(self):
		leftHeight,rightHeight=0,0
		if self.left!=None:
			leftHeight=self.left.getHeight()
		if self.right!=None:
			rightHeight=self.right.getHeight()
		return max(leftHeight,rightHeight)+1
	@staticmethod
	def createTree(preOrder,inOrder,preLeft,preRight,inLeft,inRight):
		if preLeft>preRight or inLeft>inRight:
			return None
		rootIndex=inOrder.find(preOrder[preLeft])
		root=Node(preOrder[preLeft])
		leftNum=rootIndex-inLeft
		root.left=Node.createTree(preOrder,inOrder,preLeft+1,preLeft+leftNum,inLeft,rootIndex-1)
		root.right=Node.createTree(preOrder,inOrder,preLeft+1+leftNum,preRight,rootIndex+1,inRight)
		return root

def main():
	for line in sys.stdin:
		n=int(line)
		preOrder=input()
		inOrder=input()
		root=Node.createTree(preOrder,inOrder,0,n-1,0,n-1)
		print(root.getHeight())

if __name__ == '__main__':
  main()

File: 卡码网/Js/test23.js
//23. 二叉树的高度
const rl=require("readline").createInterface({input:process.stdin});
const iter=rl[Symbol.asyncIterator]();
const readline=async ()=>(await iter.next()).value;

const out=process.stdout;

class Node{
  constructor(data){
    this.data=data;
    this.left=null;
    this.right=null;
  }
  getHeight(){
    let leftHeight=0,rightHeight=0;
    if(this.left!=null) leftHeight=this.left.getHeight();
    if(this.right!=null) rightHeight=this.right.getHeight();
    return Math.max(leftHeight,rightHeight)+1;
  }
  static createTree(preOrder,inOrder,preLeft,preRight,inLeft,inRight){
    if(preLeft>preRight||inLeft>inRight) return null;
    let rootIndex=inOrder.indexOf(preOrder[preLeft]);
    let root=new Node(preOrder[preLeft]);
    let leftNum=rootIndex-inLeft;
    root.left=Node.createTree(preOrder,inOrder,preLeft+1,preLeft+leftNum,inLeft,rootIndex-1);
    root.right=Node.createTree(preOrder,inOrder,preLeft+1+leftNum,preRight,rootIndex+1,inRight);
    return root;
  }
}

async function main(){
  while(line=await readline()){
    n=parseInt(line);
    let preOrder=await readline();
    let inOrder=await readline();
    let root=Node.createTree(preOrder,inOrder,0,n-1,0,n-1);
    console.log(root.getHeight());
  }
}

main();